### PR TITLE
Make scheduling benchmark more realistic

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

* Current schedule bench is quite unrealistic.
* Redesign it based on the architecture of a typical bevy app.

## Solution

Model the benchmarks around plugins, forming groups of systems.
* Systems within a group have orderings specified based on their `TypeId`.
* Inter-system dependencies are based on derived labels.
* Plugins have their own labels, which is passed down to all systems they contain. This allows plugin-wide dependencies sometimes.

### CODE REVIEW

File: benches/benches/bevy_ecs/scheduling/schedule.rs
Comment (alice-i-cecile) [inline benches/benches/bevy_ecs/scheduling/schedule.rs:208]: This needs to have a fixed seed.

File: benches/benches/bevy_ecs/scheduling/schedule.rs
Comment (alice-i-cecile) [inline benches/benches/bevy_ecs/scheduling/schedule.rs:117]: IMO we should define a simple macro here.

File: benches/benches/bevy_ecs/scheduling/schedule.rs
Comment (alice-i-cecile) [inline benches/benches/bevy_ecs/scheduling/schedule.rs:303]: Another simple macro.

File: benches/benches/bevy_ecs/scheduling/schedule.rs
Comment (alice-i-cecile) [inline benches/benches/bevy_ecs/scheduling/schedule.rs:95]: I would like the number of plugins and systems to be pulled out into constants here.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Love the constants, love the strategy for defining the graphs, love the increased realism. Eventually I'd be interested in a follow-up PR that adds a method to compute these key statistics on `Schedules` so we can ask users for usage data.


I think we can make this nicer with code generation macros, and strongly feel that this needs to be explicitly seeded.

Should be great!

File: benches/benches/bevy_ecs/scheduling/schedule.rs
Comment (JoJoJet) [reply benches/benches/bevy_ecs/scheduling/schedule.rs:95]: The problem with this is that, afaik, every single const generic type needs to be explicitly named somewhere. You can't loop over const generics. I can make it prettier with macros, but I don't think it's possible to extract everything into constants.

Comment (alice-i-cecile) [reply benches/benches/bevy_ecs/scheduling/schedule.rs:95]: Hmm, that's surprising 🤔 I'm not familiar enough with the details of macros to know if what I'm asking for is possible. We should see if we can get more expert opinions.

Comment (alice-i-cecile) [reply benches/benches/bevy_ecs/scheduling/schedule.rs:95]: Initial digging is promising:
1. https://stackoverflow.com/questions/67038789/writing-const-generic-enum-combinations-using-macro
2. https://stackoverflow.com/a/70233895

And bingo: https://docs.rs/seq-macro/latest/seq_macro/

**Review by alice-i-cecile — APPROVED** [review]
This has my approval now :) I'd still like to be able to extract out the number of systems / plugins into constants, but I won't block on it in this PR if you can't get it working.

**Comment by JoJoJet** [comment]
> I'd still like to be able to extract out the number of systems / plugins into constants, but I won't block on it in this PR if you can't get it working

So here's what I've gathered so far.

* `seq_macro` only works for literal consts, not const generics.
* Refactoring the code to not use const generics adds a ton of noise to the benchmarks (maybe because the compiler can't unroll the system-adding loops?).

How would you feel about moving the system-adding code out of benchmark, and *only* timing the first call to `SystemStage::run`? Originally I wanted to include `add_system()` calls in the benchmarks since that is a real part of bevy apps, but it might not be worth it.

**Comment by alice-i-cecile** [comment]
Sounds good. We can clean this up more later.

**Comment by JoJoJet** [comment]
Closing this since it's very out of date. If someone wants to update or replicate this, feel free.

**Pull request closed without merging**
